rust/crd/src/memory.rs

```diff
@@ -0,0 +1,208 @@
+    NoMemoryLimitsDefined,
+    #[snafu(display("could not derive memory distribution, no CPU limits defined"))]
+    NoCpuLimitsDefined,
+}
+
+/// This struct takes the resource limits of the Pod and derives Druid settings from it.
+/// For mentioned Druid properties, consult the
+/// [Druid Configuration Reference](https://druid.apache.org/docs/latest/configuration/index.html)
+/// for additional information.
+/// Also have a look at the "Basic Cluster Tuning" documentation:
+/// `<https://druid.apache.org/docs/latest/operations/basic-cluster-tuning.html>`
+pub struct HistoricalDerivedSettings {
+    total_memory: MemoryQuantity,
+    cpu_millis: CpuQuantity,
+    min_heap_ratio: f32,
+    max_buffer_size: MemoryQuantity,
+    os_reserved_memory: MemoryQuantity,
+}
+
+impl HistoricalDerivedSettings {
+    pub fn new(total_memory: MemoryQuantity, cpu_millis: CpuQuantity) -> Self {
+        Self {
+            total_memory,
+            cpu_millis,
+            min_heap_ratio: MIN_HEAP_RATIO,
+            os_reserved_memory: *RESERVED_OS_MEMORY,
+            max_buffer_size: *MAX_DIRECT_BUFFER_SIZE,
+        }
+    }
+
+    /// The total memory we use for druid. This is what's left after we take out the OS reserved memory.
+    pub fn allocatable_memory(&self) -> MemoryQuantity {
+        self.total_memory - self.os_reserved_memory
+    }
+
+    /// How much memory to set for the JVM to use. The minimum ratio can be defined in the struct.
+    /// Once the direct memory is maxed out, all the remaining allocatable memory will be assigned
+    /// as heap memory.
+    pub fn heap_memory(&self) -> MemoryQuantity {
+        // TODO also implement max limit of 24Gi, as recommended by Druid
+        self.allocatable_memory() - self.direct_access_memory()
+    }
+
+    /// The memory that is available to allocate for direct access.
+    pub fn allocatable_direct_access_memory(&self) -> MemoryQuantity {
+        self.allocatable_memory() * (1. - self.min_heap_ratio)
+    }
+
+    /// The max memory to allocate to direct access. This is based on the max buffer size of a single buffer.
+    pub fn max_direct_access_memory(&self) -> MemoryQuantity {
+        self.max_buffer_size * self.total_num_buffers() as f32
+    }
+
+    /// How much to allocate (or keep free) for direct access.
+    /// this is the amount to configure in the JVM as the `MaxDirectMemorySize`.
+    pub fn direct_access_memory(&self) -> MemoryQuantity {
+        if self.max_direct_access_memory() < self.allocatable_direct_access_memory() {
+            self.max_direct_access_memory()
+        } else {
+            self.allocatable_direct_access_memory()
+        }
+    }
+
+    /// The number of threads to use, based on the CPU millis.
+    /// leaves at least 500m available to core functionalities.
+    /// Druid Property: `druid.processing.numThreads`
+    pub fn num_threads(&self) -> usize {
+        (self.cpu_millis.as_cpu_count().round() - 1.).max(1.) as usize
+    }
```

Comment:
it is because of the rounding. For example:
2.6 milis -> (3-1).max(1) => 2 threads; 600m left
2.4 milis -> (2-1).max(1) => 1 thread; 1400m left

Does that help?

Indeed, if we're having less than 1500m, we don't leave >500m for the core, but that's because we don't have enough CPU millis altogether.